### ISSUE TITLE
Support XiP from PSRAM

### ISSUE THREAD

**Issue by Dominaezzz** [esp-rs/esp-hal#2083]
https://docs.espressif.com/projects/esp-idf/en/stable/esp32s3/api-guides/external-ram.html#execute-in-place-xip-from-psram

[CONFIG_SPIRAM_FETCH_INSTRUCTIONS](https://docs.espressif.com/projects/esp-idf/en/stable/esp32s3/api-reference/kconfig.html#config-spiram-fetch-instructions) and [CONFIG_SPIRAM_RODATA](https://docs.espressif.com/projects/esp-idf/en/stable/esp32s3/api-reference/kconfig.html#config-spiram-rodata) are the specific esp-idf configs.

In short, this feature lets you copy the application and data from flash to PSRAM at start up, which means flash is no longer needed for the rest of your application.

This is important for #2081 where you want to store huge buffers in PSRAM and DMA them to a peripheral, but you don't want flash access temporarily disabling psram access, leading to the DMA being starved of data and the peripheral getting garbage when it inevitably goes too far ahead.

Somewhat related #1083 .

**Comment by bjoernQ** [comment]
Turns out at least on ESP32-S3 we can execute code from PSRAM right after initializing PSRAM as we currently do - it's just that it's not writable via ibus (but we can via dbus and need to make sure to synchonize caches)

I think there are two ways

- the original request from #1083 extended by linking specific functions into PSRAM - for this we would need to create suitable sections in the linker scripts, copy code and data to PSRAM before anyone is allowed to access it. One challenge might be that currently we map PSRAM right after FLASH so the linker script needs to calculate the start address accordingly
- the way ESP-IDF does it (i.e. copy everything to PSRAM and then map PSRAM at the origin of external address space) - see https://github.com/espressif/esp-idf/blob/3c99557eeea4e0945e77aabac672fbef52294d54/components/esp_psram/mmu_psram_flash.c#L46-L134

**Comment by yanshay** [comment]
I'd like to place my vote in favor of implementing this issue and explain the rationale.

If I understand this right, this feature is practically required for applications that need to use large RGB displays (where buffer is in PSRAM) while at the same time using PSRAM intensively for application use.

Well, for complex applications, with user interface, using WiFi, SSL, async, etc. the limited amount of RAM on esp32 is quickly exhausted and PSRAM is required for memory allocations.
And for large displays (over 3.5") RGB is mandatory and required PSRAM memory for buffering.

There are many applications of this sort out there, and currently they are all written in C++ and out of the reach of rust developers.

Therefore, to enter the space of UI based applications, this feature is a must. I think that would be an effort wisely invested.

And to sum up: 
_"That's one small step for esp-hal, one giant leap for rustaceans."_

**Comment by bjoernQ** [comment]
The simplest thing to try I can think of would be:

- map PSRAM as we already do, FLASH is still mapped
- just memcopy everything from the mapped flash to psram
- map psram to the origin of extram, invalidate caches

I _think_ ESP-IDF directly reads flash / writes psram via SPI but just memcpy the mapped flash/psram should be easier (no code to port from ESP-IDF)

I haven't tried that (since then we would already have it :) ) but I guess it could work this way. All necessary changes would be in https://github.com/esp-rs/esp-hal/blob/4b66e3dba736f30ca370f670a8d3e1012fde50a3/esp-hal/src/soc/esp32s3/psram.rs#L116-L234

**Comment by EliteTK** [comment]
The recent 38C3 talk on liberating the WiFi on the ESP32 mentioned that reverse-engineering the ROM is allowed by espressif. So, after a bit of reverse engineering work and studying the IDF sources I've gotten XiP from PSRAM working on my ESP32S3.

<https://gist.github.com/EliteTK/5a409431082b4a4c34bb560243f2cf61>

The code is ported from the equivalent ESP-IDF with any excess fluff removed. The comments contain the reverse engineered, cleaned up, mostly valid C which represents what the `Cache_` functions which get called are actually doing.

I haven't actually tested this with a display yet as I never tried writing the DMA from PSRAM code mentioned in the related issues because I knew (from experience with esp-idf-hal, before I enabled XiP from PSRAM) that it would be glitchy and unusable. But I hope to do that tomorrow or next week.

I think it would make sense to put this in `init_psram` with additional PsramConfig fields for specifying that you want this feature. Putting the .text and .rodata in PSRAM has an effect on the amount of remaining available PSRAM which is another reason why I think it make sense to do this there. I plan on putting together a draft PR to this effect.

The code for doing this for the ESP32-S2 and ESP32-S3 is very similar. There's a v2 file in the IDF which handles ESP32-P4. I don't actually know enough about ESP32 to know if that's the extent of SOCs which support this feature.

In general I have minimal experience with esp-hal, limited experience with embedded rust, limited experience with unsafe rust, and only about a year of proper rust experience so please do comment on the gist on anything you think looks odd/could be improved, I'll happily incorporate it as I keep working on this.

**Comment by bjoernQ** [comment]
Wow - didn't notice there is a `Cache_Flash_To_SPIRAM_Copy` ROM function - nice

**Comment by EliteTK** [comment]
I've now opened a WIP PR for this in #3024 . I would appreciate any and all feedback.

**Comment by bjoernQ** [comment]
On ESP32-P4 this might be even more interesting to have since PSRAM is hex-SPI and can clock up to 250MHz

**Comment by bugadani** [comment]
> hex-SPI

That's a fun name for it, it's 16 lines, not 6, at which point I'm questioning SPI - it's a parallel bus with a chip select signal :D

**Comment by yanshay** [comment]
In P4 - do Flash and PSRAM still use the same SPI and that arbitration I was told about still take place?

**Comment by bugadani** [comment]
No, PSRAM uses a separate set of lines that are not routed to GPIOs.
